library/include/ck/library/utility/fill.hpp

```diff
@@ -25,6 +28,15 @@ struct FillUniformDistribution
         std::uniform_real_distribution<float> dis(a_, b_);
         std::generate(first, last, [&dis, &gen]() { return ck::type_convert<T>(dis(gen)); });
     }
+
+    template <typename ForwardRange>
+    auto operator()(ForwardRange&& range) -> std::void_t<decltype(
+        std::declval<FillUniformDistribution>()(std::begin(std::forward<ForwardRange>(range)),
+                                                std::end(std::forward<ForwardRange>(range))))>
```

Comment:
Is this deduction of `auto` return value really necessary? It seems to be a over complicated way of saying `void`...

Reply:
This is _Detection Idiom_, which is usually used in templates to check if the type parameter satisfied required (syntax) constraints (**_SFINAE_**).  Only from _C++20_, we can use _requires-clause_ to simplify this kind of declarations:  

```c++
template <typename ForwardRange>
requires requires (FillUniformDistribution& dist, ForwardRange&& range) {
    dist(std::begin(std::forward<ForwardRange>(range)), std::end(std::forward<ForwardRange>(range)));
}
auto operator()(ForwardRange&& range) {
    (*this)(std::begin(std::forward<ForwardRange>(range)), std::end(std::forward<ForwardRange>(range)));
}
```

The declaration tells reader that this method only exist if the call syntax is well-formed.

Reply:
Thanks for explanation. Now I see it's part of ranges support.

Reply:
Yep, this new method is just an adapter to the old one, so it can only accept arguments which the old one takes.